subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheAccess.java

```diff
@@ -573,9 +573,9 @@ private void checkCompatibleValueSerializer(Collection<String> faultMessages, Se
         }
 
         private void checkCompatibleCacheDecorator(Collection<String> faultMessages, CacheDecorator cacheDecorator) {
-            String requestClassName = ClassUtils.getShortCanonicalName(cacheDecorator, null);
-            String currentClassName = ClassUtils.getShortCanonicalName(parameters.getCacheDecorator(), null);
-            if (!Objects.equal(requestClassName, currentClassName)) {
+            if (!Objects.equal(cacheDecorator, parameters.getCacheDecorator())) {
+                String requestClassName = ClassUtils.getShortCanonicalName(cacheDecorator, null);
+                String currentClassName = ClassUtils.getShortCanonicalName(parameters.getCacheDecorator(), null);
                 faultMessages.add(
                     String.format(" * Requested cache decorator type (%s) doesn't match current cache type (%s)",
```

Comment:
I think it's OK to just rely on the default `toString` for the two cache decorators.  It'll work if the cache decorators are the same type, but different instances or if they are different types or if only one cache decorator is null.

If we just show class names, you won't be able to tell there are two different instances of the same class.